dataclasses.py

```diff
@@ -602,3 +603,72 @@ def _astuple_inner(obj, tuple_factory):
                           for k, v in obj.items())
     else:
         return deepcopy(obj)
+
+
+def replace(obj, **changes):
+    """Return a new object replacing specified fields with new values. This
+    is especially useful for frozen classes.  Example usage::
+
+      @dataclass(frozen=True)
+      class C:
+          x: int
+          y: int
+
+      c = C(1, 2)
+      c1 = replace(c, x=3)
+      assert c1.x == 3 and c1.y == 2
+      """
+
+    if not isdataclass(obj):
+        raise TypeError("replace() should be called on dataclass instances")
+
+    # Keep track of the non-init fields to specify.
+    non_init_fields = {}
+
+    # For each field:
+    # If it's an init field and it's not in 'changes', then get it's
+    #  value from 'obj'.  If it it's an init field and it is in
+    #  'changes', leave the new value in 'changes'.
+    # If it's not an init field, then if it's in 'changes', remember
+    #  it and remove it from 'changes'. If it's not in 'changes', use
+    #  the value in 'obj'.
+    for f in fields(obj).values():
+        if f.init:
+            if f.name not in changes:
+                changes[f.name] = getattr(obj, f.name)
+        else:
+            try:
+                # If this field is in changes, remember it's value
+                #  there and remove it from changes (since it can't
+                #  be specified when creating the new object).
+                non_init_fields[f.name] = changes[f.name]
+                del changes[f.name]
+            except KeyError:
+                # This field wasn't in changes, use the current
+                #  value of the field from obj.
+                non_init_fields[f.name] = getattr(obj, f.name)
+
+    # Create the new object, which calls __init__(), using all of the
+    #  init fields we've collected and/or left in 'changes'.
+    # If there are values supplied in changes that aren't fields, this
+    #  will correctly raise a TypeError.
+    new_obj = obj.__class__(**changes)
+
+    # Now, set fields that aren't init=True params to.
+    # For frozen objects, we have to call object.__setattr__. But we
+    #  don't know if an object is frozen or not.
+    # So the first time through, try to set the attribute, catch the
+    #  exception if one is raised, and remember if it's frozen.
+    setter = setattr
+    for idx, (name, value) in enumerate(non_init_fields.items()):
+        if idx == 0:
```

Comment:
I'm not so keen on using enumerate() just to special-case the first iteration; it's creating an extra tuple for each item. It would look more idiomatic with a flag 'first' (even though it'll take a few more lines).

Reply:
Okay, I'll do that.